pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -405,80 +408,115 @@ func (f *fqdnController) deleteRuleSelectedPods(ruleID string) error {
-			if _, ok := oldDNSMeta.responseIPs[ipStr]; !ok {
-				addressUpdate = true
-				break
+		// check for new IPs.
+		for newIPStr, newIPMeta := range newIPWithTTLMap {
+			if _, exist := cachedDNSMeta.responseIPs[newIPStr]; !exist {
+				addIPtoIPWithTTLMap(newIPStr, newIPMeta)
+				if !addressUpdate {
+					addressUpdate = true
+				}
 			}
 		}
-		for oldIPStr, oldIP := range oldDNSMeta.responseIPs {
-			if _, ok := responseIPs[oldIPStr]; !ok {
-				if oldDNSMeta.expirationTime.Before(time.Now()) {
-					// This IP entry has already expired and not seen in the latest DNS response.
-					// It should be removed from the cache.
+
+		// check for presence of already cached IPs in the new response.
+		for cachedIPStr, cachedIPMeta := range cachedDNSMeta.responseIPs {
+			if newIPMeta, exist := newIPWithTTLMap[cachedIPStr]; !exist {
+				// An already cached IP has been found in new response.
+				if cachedIPMeta.expirationTime.Before(currentTime) {
+					// this IP is expired and stale, remove it by not including it but also signal an update to syncRules.
 					addressUpdate = true
 				} else {
-					// Add the unexpired IP entry to responseIP and update the lowest applicable TTL if needed.
-					responseIPs[oldIPStr] = oldIP
-					if oldDNSMeta.expirationTime.Before(recordTTL) {
-						recordTTL = oldDNSMeta.expirationTime
-					}
+					// It hasn't expired yet, so just retain it with its existing expirationTime.
+					addIPtoIPWithTTLMap(cachedIPStr, cachedIPMeta)
 				}
+			} else {
+				// This already cached IP is part of the current response, so update it with max time between received time and its old cached time.
+				expTime := laterOf(newIPMeta.expirationTime, cachedIPMeta.expirationTime)
+				addIPtoIPWithTTLMap(cachedIPStr, ipWithTTL{
+					ip:             cachedIPMeta.ip,
+					expirationTime: expTime,
+				})
 			}
 		}
+
 	} else {
+		// First time seeing this domain.
+		// check if this needs to be tracked, by checking its presence in the datapath rules.
+		// If a FQDN policy had been applied then there must be rule records but because it's
+		// not in cache hence its FQDN:SelectorItem mapping may not be present.
+
+		// iterate over current rules mapping
+		addToCache := false
 		for selectorItem := range f.selectorItemToRuleIDs {
 			// Only track the FQDN if there is at least one fqdnSelectorItem matching it.
 			if selectorItem.matches(fqdn) {
-				mustCacheResponse, addressUpdate = true, true
+				// A FQDN can have multiple selectorItems mapped, hence we do not break the loop upon a match, but
+				// keep iterating to create mapping of multiple selectorItems against same FQDN.
+				if !addToCache {
+					addToCache = true
+				}
+
 				f.setFQDNMatchSelector(fqdn, selectorItem)
 			}
 		}
+		if addToCache {
+			for ipStr, ipMeta := range newIPWithTTLMap {
+				addIPtoIPWithTTLMap(ipStr, ipMeta)
+			}
+			addressUpdate = true
+		}
 	}
-	if mustCacheResponse {
+
+	if len(ipWithTTLMap) > 0 {
```

Comment:
I am confused, I thought we were supposed to have `addressUpdate` as the condition here?

should the code be a bit like this:
```go
if addressUpdate {
        // I think len(ipWithTTLMap) is guaranteed to be non-empty in this case, do we need to check and take a different action if it is empty?
        f.dnsEntryCache[fqdn] = ...
        f.dnsQueryQueue.AddAfter(fqdn, minTimeToRequery.Sub(currentTime))
}
```

@Dyanngg do you think `ipWithTTLMap` can be empty if `addressUpdate` is true? And if yes, do we need to take a different action, such as remove the `fqdn` key from the `f.dnsEntryCache` map?

@hkiiita the fact that this is a bug in the logic that is not caught by a unit test means we probably need an extra unit test

Reply:
@antoninbas  In this case, I'm using ```addressUpdate``` solely as a signal for ```syncDirtyRules()``` to manage any actual changes or updates to the IPs that it needs to address. 

The situation arises when a new DNS response contains only the old IPs. This means there haven't been any actual changes (no new IPs added , nor any old removed), so we don't need to set ```addressUpdate``` to ```true```. However, even though the IPs remain the same, we still need to _update_ their respective TTLs with the new values. This is why there's a distinction.

Reply:
I think you're right and I misunderstood the issue we discussed at the last meeting.
Thanks for clarifying.